torch/csrc/jit/passes/onnx/function_substitution.cpp

```diff
@@ -2,11 +2,84 @@
 
 #include <torch/csrc/jit/jit_log.h>
 #include <torch/csrc/jit/passes/onnx/helper.h>
+#include <torch/csrc/jit/passes/onnx/naming.h>
 
 namespace torch {
 namespace jit {
 
+namespace {
+
+const std::string top_module_variable_name = "";
+
+std::string TidyClassNameFromTorchScript(
+    const c10::optional<c10::QualifiedName>& class_name) {
+  if (!class_name) {
+    return "UNKNOWN_CLASS";
+  }
+  std::string out = "";
+  for (const auto& atom : class_name->atoms()) {
+    bool is_internal_torch_atom = (atom == "__torch__");
+    bool is_mangle_atom = (atom.find("__torch_mangle") != std::string::npos);
+    if (!is_internal_torch_atom && !is_mangle_atom) {
+      if (!out.empty()) {
+        out += ".";
+      }
+      out += atom;
+    }
+  }
+  return out;
+}
+
+std::string GetCallNodeVariableName(const Node* call_node) {
+  TORCH_INTERNAL_ASSERT(
+      call_node->kind() == prim::CallFunction ||
+      call_node->kind() == prim::CallMethod);
+  auto module_node = call_node->input(0)->node();
+
+  if (!module_node->hasAttribute(attr::name)) {
+    return "";
+  }
+  std::string module_name = module_node->s(attr::name);
+  if (module_node->inputs().size() == 0) {
+    return module_name;
+  }
+  // If module is from container, attr::name in module node only carries
+  // index info. Need to check parent node (container) for variable name.
+  auto parent_module_value = module_node->input(0);
+  while (parent_module_value) {
+    auto parent_module_type = parent_module_value->type()->cast<ClassType>();
+    if (parent_module_type &&
+        parent_module_type->name() ==
+            "__torch__.torch.nn.modules.container.ModuleList") {
+      auto parent_module_node = parent_module_value->node();
+      module_name = parent_module_node->s(attr::name) + "." + module_name;
+      parent_module_value = parent_module_node->inputs().size() > 0
+          ? parent_module_node->input(0)
+          : nullptr;
+    } else {
+      break;
+    }
+  }
+
+  return module_name;
+}
+
+ScopePtr ForwardCallScope(Graph& graph, Node* call_node) {
+  TORCH_INTERNAL_ASSERT(call_node->kind() == prim::CallMethod);
+  const std::string& method_name = call_node->s(attr::name);
+  if (method_name == "forward") {
+    const auto type = call_node->input(0)->type()->expect<c10::NamedType>();
+    const auto class_name = TidyClassNameFromTorchScript(type->name());
+    const auto variable_name = GetCallNodeVariableName(call_node);
+    const auto scope_name =
```

Comment:
Perhaps use explicit types for these variables for clarity?

Reply:
Done for the 3 string cases. I think they are fine either way. It wouldn't hurt readability too much, while there weren't much convenience offered by auto, since I don't expect the function signature and return type to change in the future.
Keeping `type` as is, since its type is explicitly spelled out in `expect` later.